### PR TITLE
Ranges support

### PR DESCRIPTION
**Pull request by ghost**
As discussed in #60 , I made the necessary changes to add some basic support for specific ranges in requests (RFC 7233).

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -627,6 +627,12 @@ pub fn ErrorPreconditionFailed<T>(err: T) -> InternalError<T> {
     InternalError::new(err, StatusCode::PRECONDITION_FAILED)
 }
 
+///  Helper function that creates wrapper of any error and generate *RANGE NOT SATISFIABLE* response.
```
Comment (fafhrd91) [inline src/error.rs]: i dont think we need this one. this error is too specific

File: src/fs.rs [hunk 2]
```diff
@@ -83,16 +84,37 @@ impl Responder for NamedFile {
...
     type Error = io::Error;
 
-    fn respond_to(mut self, _: HttpRequest) -> Result<HttpResponse, io::Error> {
-        let mut resp = HTTPOk.build();
-        resp.content_encoding(ContentEncoding::Identity);
-        if let Some(ext) = self.path().extension() {
+    fn respond_to(mut self, req: HttpRequest) -> Result<HttpResponse, io::Error> {
+        if let Some(rangeheader) = req.headers().get(header::RANGE) {
+            let file_metadata = metadata(self.0)?;
+            if let Ok(ranges) = HttpRange::parse(rangeheader.to_str().unwrap(), file_metadata.len()) {
```
Comment (fafhrd91) [inline src/fs.rs]: you should not "unwrap" here, in case of error you should return `HTTPRangeNotSatisfiable`

File: src/fs.rs [hunk 3]
```diff
@@ -83,16 +84,37 @@ impl Responder for NamedFile {
...
+                let mut resp = HTTPPartialContent.build();
+                let length: usize = ranges[0].length as usize;
+                let mut data: Vec<u8> = vec![0u8; length];
+                let _ = &self.1.seek(SeekFrom::Start(ranges[0].start))?;
+                let _ = self.1.read_exact(&mut data);
```
Comment (fafhrd91) [inline src/fs.rs]: you should use `?`

File: src/fs.rs [hunk 4]
```diff
@@ -83,16 +84,37 @@ impl Responder for NamedFile {
...
+                Ok(resp
+                    .header(header::CONTENT_RANGE, format!("bytes {}-{}/{}", ranges[0].start, ranges[0].length, file_metadata.len()).as_str())
+                    .body(data).unwrap())
+            } else {
+                Ok(HTTPRangeNotSatisfiable.build()
+                    .header(header::CONTENT_RANGE, format!("bytes */{}", file_metadata.len()).as_str())
+                    .header(header::ACCEPT_RANGES, "bytes")
+                    .body("").unwrap())
```
Comment (fafhrd91) [inline src/fs.rs]: use `.finish().unwrap()`

File: src/fs.rs [hunk 5]
```diff
@@ -83,16 +84,37 @@ impl Responder for NamedFile {
...
+            }
+        } else {
+            //let file_metadata = metadata(self.0)?;
```
Comment (fafhrd91) [inline src/fs.rs]: remove this

**Review by fafhrd91 — CHANGES_REQUESTED** [review]
Could you also rebase your PR

**Comment by ghost** [comment]
I close this for now, since I noticed some other bugs in the code, also I have to get a proper mergetool. I'll reopen if I finally ironed this out.

**Pull request closed without merging**